internal/controller/collection/reconciler.go

```diff
@@ -0,0 +1,277 @@
+	errStatusUpdate        = "cannot update status"
+	fieldOwner             = client.FieldOwner("observed-object-collection")
+)
+
+// Reconciler watches for ObservedObjectCollection resources
+// and creates observe-only Objects for the matched items.
+type Reconciler struct {
+	client             client.Client
+	log                logging.Logger
+	pollInterval       func() time.Duration
+	clientForProvider  func(ctx context.Context, inclusterClient client.Client, providerConfigName string) (client.Client, error)
+	observedObjectName func(collection client.Object, matchedObject client.Object) (string, error)
+}
+
+// Setup adds a controller that reconciles ObservedObjectCollection resources.
+func Setup(mgr ctrl.Manager, o controller.Options, pollJitter time.Duration) error {
+	name := managed.ControllerName(v1alpha1.ObservedObjectCollectionGroupKind)
+
+	r := &Reconciler{
+		client: mgr.GetClient(),
+		log:    o.Logger,
+		pollInterval: func() time.Duration {
+			return o.PollInterval + +time.Duration((rand.Float64()-0.5)*2*float64(pollJitter)) //nolint
+		},
+		clientForProvider:  clients.ClientForProvider,
+		observedObjectName: observedObjectName,
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.ObservedObjectCollection{}).
+		WithEventFilter(predicate.Or(
+			predicate.GenerationChangedPredicate{},
+			predicate.AnnotationChangedPredicate{},
+			predicate.LabelChangedPredicate{}),
+		).
+		Complete(ratelimiter.NewReconciler(name, xperrors.WithSilentRequeueOnConflict(r), o.GlobalRateLimiter))
+}
+
+// Reconcile fetches objects specified by their GVK and label selector
+// and creates observed-only Objects for the matches.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, error error) { //nolint:gocyclo
+	log := r.log.WithValues("request", req)
+
+	defer func() {
+		if error == nil {
+			log.Info("Reconciled")
+		} else {
+			log.Info("Retry", "err", error)
+		}
+	}()
+
+	collection := &v1alpha1.ObservedObjectCollection{}
+	err := r.client.Get(ctx, req.NamespacedName, collection)
+
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	if meta.WasDeleted(collection) {
+		return ctrl.Result{}, nil
+	}
+
+	if meta.IsPaused(collection) {
+		collection.Status.SetConditions(xpv1.ReconcilePaused())
+		return ctrl.Result{}, r.client.Status().Update(ctx, collection)
+	}
+
+	log.Info("Reconciling", "name", collection.Name)
+
+	// Get client for the referenced provider config.
+	clusterClient, err := r.clientForProvider(ctx, r.client, collection.Spec.ProviderConfigReference.Name)
+	if err != nil {
+		collection.Status.SetConditions(xpv1.ReconcileError(err))
+		if err := r.client.Status().Update(ctx, collection); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+		}
+		return ctrl.Result{}, errors.Wrap(err, errNewKubernetesClient)
```

Comment:
Looks like we are loosing the error from `r.clientForProvider` here. Overridden by the result of `r.client.Status().Update`.

Reply:
Fixed, please resolve it it matches the expectations.